data_model/src/block.rs

```diff
@@ -193,76 +219,85 @@ impl SignedBlock {
         iroha_crypto::HashOf::new(self)
     }
 
-    /// Calculate block payload [`Hash`](`iroha_crypto::HashOf`).
-    #[inline]
-    #[cfg(feature = "std")]
-    #[cfg(feature = "transparent_api")]
-    pub fn hash_of_payload(&self) -> iroha_crypto::HashOf<BlockPayload> {
-        let SignedBlock::V1(block) = self;
-        iroha_crypto::HashOf::new(&block.payload)
-    }
-
-    /// Add additional signatures to this block
-    #[must_use]
-    #[cfg(all(feature = "std", feature = "transparent_api"))]
-    pub fn sign(mut self, key_pair: &iroha_crypto::KeyPair) -> Self {
-        let SignedBlock::V1(block) = &mut self;
-        let signature = iroha_crypto::SignatureOf::new(key_pair, &block.payload);
-        block.signatures.insert(signature);
-        self
-    }
-
-    /// Add additional signatures to this block
+    /// Add signature to the block
     ///
     /// # Errors
     ///
-    /// If given signature doesn't match block hash
+    /// if signature is invalid
     #[cfg(feature = "transparent_api")]
     pub fn add_signature(
         &mut self,
-        signature: iroha_crypto::SignatureOf<BlockPayload>,
-    ) -> Result<(), iroha_crypto::error::Error> {
-        let SignedBlock::V1(block) = self;
-        signature.verify(&block.payload)?;
+        signature: BlockSignature,
+        public_key: &iroha_crypto::PublicKey,
+    ) -> Result<(), iroha_crypto::Error> {
+        #[cfg(not(feature = "std"))]
+        use alloc::{borrow::ToOwned as _, collections::BTreeSet};
+        #[cfg(feature = "std")]
+        use std::collections::BTreeSet;
+
+        if !self
+            .signatures()
+            .collect::<BTreeSet<_>>()
+            .insert(&signature)
+        {
+            return Err(iroha_crypto::Error::Signing(
+                "Duplicate signature".to_owned(),
+            ));
+        }
```

Comment:
Afaik we compare `BlockSingature` by both fields `index` and `signature`, but some siganture schemes can produce multiple valid results for the same output. 
So malicious peer still can produce multiple votes for a block.
Can we compare just by index?

Reply:
I agree with comparing only index in `BlockSignature`